cdap-docs/developers-manual/source/building-blocks/datasets/overview.rst

```diff
@@ -152,6 +152,26 @@ Similarly to static datasets, if a program is multi-threaded, CDAP will make
 sure that every thread has its own instance of each dynamic dataset |---| and in order to discard a dataset
 from the cache, every thread that uses it must individually call ``discardDataset()``.
 
+.. _multi-threading-dataset-access:
+
+Multi-threading and Dataset Access
+----------------------------------
+
+As mentioned above, if a program is is multi-threaded, CDAP will make sure that every
+thread has its own instance of a dataset. This is because datasets are not thread-safe,
+and each thread must operate on its own instance of a Dataset.
+
+As a consequence, multiple threads accessing the same dataset will have different
+instances of the same dataset. If they are in the same transaction, an issue arises if the
+datasets are :ref:`Table-based <datasets-overview-types>`: the threads will buffer writes
+in-memory and will not see each other's writes. Then, upon flush, it is possible that the
+different threads will overwrite each others' changes.
+
+One solution is to design your use of the datasets so that each thread writes to unique
+columns, preventing the overwriting. Another solution is to use a :ref:`dataset type
+<datasets-overview-types>` that is not Table-based, such as a :ref:`FileSet
```

Comment:
I don't think we should suggest FileSet for multithreading, because what if multiple threads write to the same file? (similar as the original problem of writing to the same column of a Table)